internal/eval/deque.go

```diff
@@ -215,6 +302,179 @@ func (q *Deque) RPop() (string, error) {
+		}
+	}
+}
+
+func (q *Deque) insertAfterPivotNode(element string, qIterator *DequeIterator, pivotNode *byteListNode) {
+	if qIterator.ElementsTraversed == q.Length {
+		// Element needs to be inserted at the end of the Deque.
+		q.RPush(element)
+	} else {
+		// Element needs to be inserted b/w 2 nodes in the Deque.
+		q.insertAfterPivotNodeHelper(element, qIterator, pivotNode)
+	}
+}
+
+func (q *Deque) getNewNodeWithElement(element string) *byteListNode {
+	elementEntryLen := int(GetEncodeDeqEntrySize(element))
+	elementNode := q.list.newNodeWithCapacity(elementEntryLen)
+	elementNode.buf = elementNode.buf[:elementEntryLen]
+	EncodeDeqEntryInPlace(element, elementNode.buf[:elementEntryLen])
+	return elementNode
+}
+
+func (q *Deque) breakPivotNodeAndInsertBefore(qIterator *DequeIterator, pivotNode, elementNode *byteListNode, pivotEntryLen, leftIdx int) *byteListNode {
+	bufIndex := qIterator.BufIndex
+	if bufIndex == 0 {
+		bufIndex = len(pivotNode.buf)
+	}
+	newNode := q.list.newNodeWithCapacity(bufIndex - pivotEntryLen - leftIdx)
+	newNode.buf = append([]byte{}, pivotNode.buf[leftIdx:bufIndex-pivotEntryLen]...)
+	pivotNode.buf = pivotNode.buf[bufIndex-pivotEntryLen:]
+	if pivotNode.prev != nil {
+		pivotNode.prev.next = newNode
+	}
+	newNode.prev = pivotNode.prev
+	newNode.next = elementNode
+	elementNode.prev = newNode
+	elementNode.next = pivotNode
+	pivotNode.prev = elementNode
+	return newNode
+}
+
+func (q *Deque) updateHeadLInsert(newNode, prevHead *byteListNode, prevLeftIdx int) {
+	if newNode.prev == nil {
+		q.list.head = newNode
+		q.leftIdx = 0
+	} else {
+		q.list.head = prevHead
+		q.leftIdx = prevLeftIdx
+	}
+}
+
+func (q *Deque) insertBeforePivotHelper(pivot, element string, qIterator *DequeIterator, pivotNode *byteListNode) {
+	pivotEntryLen := int(GetEncodeDeqEntrySize(pivot))
+	prevHead := q.list.head
+	prevLeftIdx := q.leftIdx
+	leftIdx := q.leftIdx
+	if pivotNode.prev != nil {
+		leftIdx = 0
+	}
+	elementNode := q.getNewNodeWithElement(element)
+	if qIterator.BufIndex == pivotEntryLen || (qIterator.BufIndex == 0 && (len(pivotNode.buf) == pivotEntryLen)) {
+		// No need to break the pivotNode into two nodes when the pivot element is the first element in the buffer.
+		prevNode := pivotNode.prev
+		prevNode.next = elementNode
+		pivotNode.prev = elementNode
+		elementNode.next = pivotNode
+		elementNode.prev = prevNode
+	} else {
+		newNode := q.breakPivotNodeAndInsertBefore(qIterator, pivotNode, elementNode, pivotEntryLen, leftIdx)
+		q.updateHeadLInsert(newNode, prevHead, prevLeftIdx)
+	}
+	q.Length++
+}
+
+func (q *Deque) insertBeforePivotNode(pivot, element string, qIterator *DequeIterator, pivotNode *byteListNode) {
+	if qIterator.ElementsTraversed == 1 {
+		// Element needs to be inserted at the front of the Deque.
+		q.LPush(element)
+	} else {
+		q.insertBeforePivotHelper(pivot, element, qIterator, pivotNode)
+	}
+}
+
+func (q *Deque) LInsert(pivot, element, beforeAfter string) (int64, error) {
```

Comment:
Add a comment explaining what this method does.

Reply:
Done.